### PR TITLE
Gpu swap prefetch

### PR DESCRIPTION
**Pull request by KarlLeell**

### CODE REVIEW

File: include/mxnet/gpu_swap_prefetch.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  std::vector<int> lookahead_pos = std::vector<int>(MemHistory::NUMBER_OF_GPU);
+
+  ~Prefetch();
+  static Prefetch* Get();
+  static std::shared_ptr<Prefetch> _GetSharedRef();
+  void StartPrefetching(int device);
+
+private:
+  Prefetch();
+  bool begin_prefetching_;
```
Comment (fegin) [inline include/mxnet/gpu_swap_prefetch.h]: "start" seems to be a better word than "begin". "start/stop".

File: src/storage/gpu_swap_prefetch.cc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#include <dmlc/logging.h>
+#include <mxnet/gpu_swap_history.h>
+#include <mxnet/swap.h>
+#include <mxnet/gpu_swap_prefetch.h>
+
+
+namespace mxnet {
+
+Prefetch::Prefetch() {
+  begin_prefetching__ = false;
```
Comment (fegin) [inline src/storage/gpu_swap_prefetch.cc]: Two "_"? Typo?

File: src/storage/gpu_swap_prefetch.cc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+std::shared_ptr<Prefetch> Prefetch::_GetSharedRef() {
+  static std::shared_ptr<Prefetch> inst(new Prefetch());
+  return inst;
+}
+
+
+void Prefetch::StartPrefetching(int device) {
+  pthread_t this_thread = pthread_self();
+  while(stop_prefetching_) {
```
Comment (fegin) [inline src/storage/gpu_swap_prefetch.cc]: It seems that using one variable is enough, e.g., "is_running". But you may have a good reason to use two variables. If not, change to using one variable.

**Review by fegin — CHANGES_REQUESTED** [review]
Feel free to merge after you fix my reviews.

File: include/mxnet/gpu_swap_history.h [hunk 4]
```diff
@@ -19,12 +19,12 @@ using handle_id_t = unsigned long long;
 using timestamp_t = unsigned long long;
 using timestep_t = unsigned long long;
 
-const int NUMBER_OF_GPU = 8;
 
 class MemHistory {
 public:
 
   enum record_t {GET_ADDR, SET_ADDR, DEL_ADDR};
+  static const int NUMBER_OF_GPU = 8;
```
Comment (Sotskin) [inline include/mxnet/gpu_swap_history.h]: Why has number of gpu become static variable in MemHistory?

**Review by fegin — APPROVED** [review]

**Comment by fegin** [comment]
Will do another code review sometime on Saturday.

File: include/mxnet/gpu_swap_prefetch.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  void StopPrefetching();
+  void HistoryBasedPrefetch(int device);
+  bool IsPrefetching() {return start_prefetching_;}
+
+private:
+
+  Prefetch();
+  void Prefetching(int device);
+
+  std::vector<int> lookahead_pos = std::vector<int>(MemHistory::NUMBER_OF_GPU);
```
Comment (Sotskin) [inline include/mxnet/gpu_swap_prefetch.h]: rename to lookahead_pos_

**Review by Sotskin — CHANGES_REQUESTED** [review]
Just some style changes. Looks good

**Review by qingsenzhang — APPROVED** [review]

**Review by fegin — APPROVED** [review]

**Review by Sotskin — APPROVED** [review]

**Pull request merged**